File: Wire-iOS/Sources/UserInterface/Overlay/NetworkStatus/NetworkStatusViewController.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "NetworkStatusViewController.h"

@import PureLayout;

// ui
#import "PassthroughTouchesView.h"

// helpers

#import "WAZUIMagicIOS.h"
#import "WireSyncEngine+iOS.h"
#import "UIAlertController+Wire.h"
#import "Wire-Swift.h"

typedef NS_ENUM(NSInteger, StatusBarState) {
    StatusBarStateOk,
    StatusBarStateServerUnreachable
};

typedef NS_ENUM(NSInteger, NetworkAlertType) {
    NetworkAlertTypeOffline,
    NetworkAlertTypeLowQuality
};


static CGFloat expandedHeight, collapsedHeight;
static UIFont *font;
static UIColor *fontColor, *warningBackgroundColor;



@interface NetworkStatusViewController (Animations)

- (void)expandBarWithTimeout;
- (void)collapseBarWithCompletion:(dispatch_block_t)completion;

@end



@interface NetworkStatusViewController ()

@property (nonatomic, readonly) StatusBarState statusBarState;
@property (nonatomic, assign) StatusBarState pendingStatusBarState;

@property (nonatomic, assign) BOOL statusBarIsExpanded;

@property (nonatomic, assign) BOOL monitoringEnabled;

/// The container view that is actually doing the presentation
@property (nonatomic, strong) UIView *animatedContainer;
@property (nonatomic, strong) UILabel *statusLabel;
@property (nonatomic, strong) NSLayoutConstraint *animatedContainerHeight;

@property (nonatomic, strong) UITapGestureRecognizer *tapGestureRecognizer;

@property (nonatomic, strong) NSTimer *collapseTimer;
@property (nonatomic, strong) id serverConnectionObserverToken;

@end


@interface NetworkStatusViewController (ServerConnection) <ServerConnectionObserver>

@end



@implementation NetworkStatusViewController

- (id)init
{
    self = [super init];
    if (self) {
        // Custom initialization
        _statusBarState = StatusBarStateOk;
        self.monitoringEnabled = NO;
    }
    return self;
}

- (void)loadView
{
    [super loadView];
    self.view.backgroundColor = [UIColor clearColor];
    
    // set up the container that will be animating
    
    self.statusLabel = [[UILabel alloc] initForAutoLayout];
    self.statusLabel.textAlignment = NSTextAlignmentCenter;
    
    self.animatedContainer = [[UIView alloc] initForAutoLayout];
    [self.view addSubview:self.animatedContainer];
    
    [self.animatedContainer autoPinEdgesToSuperviewEdgesWithInsets:UIEdgeInsetsZero];
    
    self.animatedContainerHeight = [self.animatedContainer autoSetDimension:ALDimensionHeight toSize:0];
    
    [self.animatedContainer addSubview:self.statusLabel];
    
    [self.statusLabel autoPinEdgesToSuperviewEdgesWithInsets:UIEdgeInsetsZero];
}

- (void)viewDidLoad
{
    [super viewDidLoad];

    expandedHeight = [WAZUIMagic floatForIdentifier:@"system_status_bar.expanded_height"];
    collapsedHeight = [WAZUIMagic floatForIdentifier:@"system_status_bar.collapsed_height"];
    font = [UIFont fontWithMagicIdentifier:@"system_status_bar.warning_font"];
    fontColor = [UIColor colorWithMagicIdentifier:@"system_status_bar.warning_font_color"];
    warningBackgroundColor = [UIColor colorWithMagicIdentifier:@"system_status_bar.warning_background_color"];
    
    self.statusLabel.font = font;
    self.statusLabel.textColor = fontColor;
    self.statusLabel.backgroundColor = [UIColor clearColor];

    [self addTapGestureRecognizer];
    
    self.serverConnectionObserverToken = [SessionManager.shared.serverConnection addServerConnectionObserver:self];
    
    // Set the view hidden to disable receiving touches. It will be eventually set to NO by a reachability change.
    self.view.hidden = YES;

    __weak NetworkStatusViewController *weakSelf = self;
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t) ([WAZUIMagic floatForIdentifier:@"system_status_bar.startup_delay"] * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        weakSelf.monitoringEnabled = YES;
        [self updateState];
    });
}

- (void)addTapGestureRecognizer
{
    if (! self.tapGestureRecognizer) {
        self.tapGestureRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(tappedOnBackground:)];
        [self.animatedContainer addGestureRecognizer:self.tapGestureRecognizer];
    }
}

- (void)tappedOnBackground:(UIGestureRecognizer *)sender
{
    if (self.animatedContainerHeight.constant == expandedHeight) {
        // if tapping on the bar, pick the type of alert to show, based on server reachability
        [self showAlertWithType:NetworkAlertTypeOffline];
    }
    else {
        [self expandBarWithTimeout];
    }
}

- (void)updateState
{
    BOOL isOffline = SessionManager.shared.serverConnection.isOffline;
    [self enqueueStatusBarStateChange:isOffline ? StatusBarStateServerUnreachable : StatusBarStateOk];
}

- (void)flashNetworkStatusIfNecessaryAndShowAlert:(BOOL)showAlert
{
    if (self.statusBarState == StatusBarStateServerUnreachable) {
        [self expandBarWithTimeout];
        
        if (showAlert) {
            [self showAlertWithType:NetworkAlertTypeOffline];
        }
    }
}

- (void)showAlertWithType:(NetworkAlertType)alertType
{
    NSString *title = NSLocalizedString(@"system_status_bar.no_internet.title", @"");
    NSString *explanation = NSLocalizedString(@"system_status_bar.no_internet.explanation", @"");
    
    if (alertType == NetworkAlertTypeLowQuality) {
        title = NSLocalizedString(@"system_status_bar.poor_connectivity.title", @"");
        explanation = NSLocalizedString(@"system_status_bar.poor_connectivity.explanation", @"");
    }
    
    NSString *acknowledge = NSLocalizedString(@"general.confirm", @"OK");
    
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:title
                                                                   message:explanation
                                                         cancelButtonTitle:acknowledge];
    [alert presentTopmost];
}

- (void)enqueueStatusBarStateChange:(StatusBarState)state
{
    if (! self.monitoringEnabled) {
        return;
    }

    // represented state is already correct, nothing to do
    if ((state == self.statusBarState) && self.pendingStatusBarState == state) {
        return;
    }
    
    self.pendingStatusBarState = state;

    [self dequeueAndRunStatusBarChange];
}

- (void)dequeueAndRunStatusBarChange
{
    if (self.pendingStatusBarState == self.statusBarState) {
        return;
    }

    _statusBarState = self.pendingStatusBarState;

    if (self.statusBarState == StatusBarStateOk) {
        
        if (self.statusBarIsExpanded) {
            
        }
        self.animatedContainerHeight.constant = 0;
        [self.view setNeedsUpdateConstraints];
        
        [self collapseBarWithCompletion:^{
            if (self.pendingStatusBarState != self.statusBarState) {
                [self dequeueAndRunStatusBarChange];
            } else {
                self.view.hidden = YES;
            }
        }];
    } else if (self.statusBarState == StatusBarStateServerUnreachable) {
        self.view.hidden = NO;
        [self expandBarWithTimeout];
    }
}

- (NSString *)titleForConnectivity
{
    return NSLocalizedString(@"system_status_bar.no_internet.title", @"");
}

- (NSString *)explanationForConnectivity
{
    return NSLocalizedString(@"system_status_bar.no_internet.explanation", @"");
}

@end



@implementation NetworkStatusViewController (ServerConnection)

- (void)serverConnectionDidChange:(id<ServerConnection>)serverConnection
{
    [self updateState];
}

@end



@implementation NetworkStatusViewController (Animations)

- (void)updateCollapseTimer
{
    if (self.collapseTimer != nil) {
        [self.collapseTimer invalidate];
    }
    NSTimeInterval timeout = [WAZUIMagic floatForIdentifier:@"system_status_bar.collapse_animation.delay_after_expand"];
    self.collapseTimer = [NSTimer scheduledTimerWithTimeInterval:timeout target:self selector:@selector(collapseBarAfterTimeout) userInfo:nil repeats:NO];
    self.collapseTimer.tolerance = 0.2;
}

- (void)expandBarWithTimeout
{
    
    if (! self.statusBarIsExpanded) {
        self.statusBarIsExpanded = YES;
        self.statusLabel.text = [[self titleForConnectivity] uppercasedWithCurrentLocale];
        self.statusLabel.alpha = 0;
        self.animatedContainer.alpha = 0;
        self.animatedContainer.backgroundColor = warningBackgroundColor;
        self.animatedContainerHeight.constant = expandedHeight;
        
        [UIView animateWithAnimationIdentifier:@"system_status_bar.expand_animation" animations:^{
            [self.view layoutIfNeeded];
            self.animatedContainer.alpha = 1;
            self.statusLabel.alpha = 1;
        } options:UIViewAnimationOptionAllowUserInteraction | UIViewAnimationOptionBeginFromCurrentState completion:^(BOOL finished) {
            [self updateCollapseTimer];
        }];
    }
    else {
        [self updateCollapseTimer];
    }
}

- (void)collapseBarAfterTimeout
{
    [self collapseBarWithCompletion:^{
        if (self.pendingStatusBarState != self.statusBarState) {
            [self dequeueAndRunStatusBarChange];
        }
    }];
}

- (void)collapseBarWithCompletion:(dispatch_block_t)completion
{
    if (self.statusBarIsExpanded) {
        self.statusBarIsExpanded = NO;
        [self.collapseTimer invalidate];
        self.collapseTimer = nil;
        self.animatedContainerHeight.constant = collapsedHeight;
        [self.view setNeedsUpdateConstraints];

        [UIView animateWithAnimationIdentifier:@"system_status_bar.collapse_animation" animations:^{
            
            self.statusLabel.alpha = 0;
            [self.view layoutIfNeeded];
            
        } options:UIViewAnimationOptionAllowUserInteraction | UIViewAnimationOptionBeginFromCurrentState completion:^(BOOL finished) {
            if (completion) {
                completion();
            }
        }];
    }
    else {
        if (self.pendingStatusBarState != self.statusBarState) {
            [self dequeueAndRunStatusBarChange];
        }
    }
}

@end

File: Wire-iOS/Sources/UserInterface/Overlay/NetworkStatus/NetworkStatusViewController.swift
//
// Wire
// Copyright (C) 2017 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import Foundation
import Cartography

typealias NetworkStatusBarDelegate = NetworkStatusViewControllerDelegate & NetworkStatusViewDelegate

protocol NetworkStatusViewControllerDelegate: class {
    /// if return false, NetworkStatusViewController will not disapper in iPad regular mode landscape orientation.
    var showInIPadLandscapeMode: Bool {get}

    /// if return false, NetworkStatusViewController will not disapper in iPad regular mode portrait orientation.
    var showInIPadPortraitMode: Bool {get}
}

extension Notification.Name {
    static let ShowNetworkStatusBar = Notification.Name("ShowNetworkStatusBar")
}

@objc
class NetworkStatusViewController : UIViewController {

    public weak var delegate: NetworkStatusBarDelegate? {
        didSet {
            networkStatusView.delegate = delegate
        }
    }

    let networkStatusView = NetworkStatusView()
    fileprivate var networkStatusObserverToken : Any?
    fileprivate var pendingState : NetworkStatusViewState?
    var state: NetworkStatusViewState?
    fileprivate var offlineBarTimer : Timer?
    fileprivate var device: DeviceProtocol = UIDevice.current

    override func loadView() {
        let passthroughTouchesView = PassthroughTouchesView()
        passthroughTouchesView.clipsToBounds = true
        self.view = passthroughTouchesView
    }

    /// default init method with a parameter for injecting mock device
    ///
    /// - Parameter device: Provide this param for testing only
    init(device: DeviceProtocol = UIDevice.current) {
        super.init(nibName: nil, bundle: nil)

        self.device = device

        NotificationCenter.default.addObserver(self, selector: #selector(changeStateFormOfflineCollapsedToOfflineExpanded), name: Notification.Name.ShowNetworkStatusBar, object: .none)

    }

    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    deinit {
        NSObject.cancelPreviousPerformRequests(withTarget: self, selector: #selector(applyPendingState), object: nil)
        NotificationCenter.default.removeObserver(self)

        offlineBarTimer?.invalidate()
        offlineBarTimer = nil
    }
    
    override func viewDidLoad() {
        view.addSubview(networkStatusView)
        
        constrain(self.view, networkStatusView) { containerView, networkStatusView in
            networkStatusView.left == containerView.left
            networkStatusView.right == containerView.right
            networkStatusView.top == containerView.top
        }

        if let userSession = ZMUserSession.shared() {
            update(state: viewState(from: userSession.networkState))
            networkStatusObserverToken = ZMNetworkAvailabilityChangeNotification.addNetworkAvailabilityObserver(self, userSession: userSession)
        }
        
        networkStatusView.addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(tappedOnNetworkStatusBar)))
    }

    func changeStateFormOfflineCollapsedToOfflineExpanded() {
        let networkStatusView = self.networkStatusView

        if networkStatusView.state == .offlineCollapsed {
            self.update(state: .offlineExpanded)
        }
    }


    /// show NetworkStatusViewController instance(s) if its state is .offlineCollapsed
    static public func notifyWhenOffline() {
        NotificationCenter.default.post(name: .ShowNetworkStatusBar, object: self)
    }

    func showOfflineAlert() {
        let offlineAlert = UIAlertController.init(title: "system_status_bar.no_internet.title".localized,
                                                  message: "system_status_bar.no_internet.explanation".localized,
                                                  cancelButtonTitle: "general.confirm".localized)
        
        offlineAlert.presentTopmost()
    }
    
    fileprivate func viewState(from networkState : ZMNetworkState) -> NetworkStatusViewState {
        switch networkState {
        case .offline:
            return .offlineExpanded
        case .online:
            return .online
        case .onlineSynchronizing:
            return .onlineSynchronizing
        }
    }
    
    internal func tappedOnNetworkStatusBar() {
        switch networkStatusView.state {
        case .offlineCollapsed:
            update(state: .offlineExpanded)
        case .offlineExpanded:
            showOfflineAlert()
        default:
            break
        }
    }
    
    fileprivate func startOfflineBarTimer() {
        offlineBarTimer = .allVersionCompatibleScheduledTimer(withTimeInterval: 2.0, repeats: false) { [weak self] _ in
            self?.collapseOfflineBar()
        }
    }
    
    internal func collapseOfflineBar() {
        if networkStatusView.state == .offlineExpanded {
            update(state: .offlineCollapsed)
        }
    }
    
    fileprivate func enqueue(state: NetworkStatusViewState) {
        pendingState = state
        NSObject.cancelPreviousPerformRequests(withTarget: self, selector: #selector(applyPendingState), object: nil)
        perform(#selector(applyPendingState), with: nil, afterDelay: 1)
    }
    
    internal func applyPendingState() {
        guard let state = pendingState else { return }
        update(state: state)
        pendingState = nil
    }
    
    func update(state : NetworkStatusViewState) {
        self.state = state

        networkStatusView.update(state: state, animated: true)
        
        if state == .offlineExpanded {
            startOfflineBarTimer()
        }
    }

}

extension NetworkStatusViewController : ZMNetworkAvailabilityObserver {
    
    func didChangeAvailability(newState: ZMNetworkState) {
        enqueue(state: viewState(from: newState))
    }
    
}

// MARK: - iPad size class and orientation switching

extension NetworkStatusViewController {
    
    func shouldShowOnIPad(for newOrientation: UIDeviceOrientation?) -> Bool {
        guard isIPadRegular(device: device) else { return true }

        guard let delegate = self.delegate, let newOrientation = newOrientation else { return true }

        if newOrientation.isPortrait {
            return delegate.showInIPadPortraitMode
        } else if newOrientation.isLandscape {
            return delegate.showInIPadLandscapeMode
        } else {
            return true
        }
    }

    func updateStateForIPad(for newOrientation: UIDeviceOrientation?) {
        if shouldShowOnIPad(for: newOrientation) {
            if let state = state {
                networkStatusView.update(state: state, animated: false)
            }
        } else {
            /// when size class changes and delegate view controller disabled to show networkStatusView, hide the networkStatusView
            networkStatusView.update(state: .online, animated: false)
        }
    }

    open override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
        super.traitCollectionDidChange(previousTraitCollection)
        guard device.userInterfaceIdiom == .pad else { return }

        updateStateForIPad(for: device.orientation)
    }

    override func viewWillTransition(to size: CGSize, with coordinator: UIViewControllerTransitionCoordinator?) {
        if let coordinator = coordinator {
            super.viewWillTransition(to: size, with: coordinator)
        }

        guard isIPadRegular(device: device) else { return }

        // find out the new orientation with the new size
        var newOrientation: UIDeviceOrientation = .unknown
        if size.width > 0 {
            if size.width > size.height {
                newOrientation =  .landscapeLeft
            } else if size.width < size.height {
                newOrientation =  .portrait
            }
        }

        updateStateForIPad(for: newOrientation)
    }

}

